File: sdk/go/import_export.go
package liteboxd

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"mime/multipart"
	"strconv"
)

// ImportExportService handles template import/export operations.
type ImportExportService struct {
	client *Client
}

// ImportTemplates imports templates from YAML.
// strategy: "create-only", "update-only", or "create-or-update"
// autoPrepull: whether to trigger prepull after import
func (ie *ImportExportService) ImportTemplates(ctx context.Context, yamlContent []byte, strategy ImportStrategy, autoPrepull bool) (*ImportTemplatesResponse, error) {
	// Create multipart form body
	var body bytes.Buffer
	writer := multipart.NewWriter(&body)

	// Add file field
	part, err := writer.CreateFormFile("file", "templates.yaml")
	if err != nil {
		return nil, fmt.Errorf("failed to create form file: %w", err)
	}
	if _, err := part.Write(yamlContent); err != nil {
		return nil, fmt.Errorf("failed to write file content: %w", err)
	}

	// Add strategy field
	if err := writer.WriteField("strategy", string(strategy)); err != nil {
		return nil, fmt.Errorf("failed to write strategy field: %w", err)
	}

	// Add prepull field
	if autoPrepull {
		if err := writer.WriteField("prepull", strconv.FormatBool(autoPrepull)); err != nil {
			return nil, fmt.Errorf("failed to write prepull field: %w", err)
		}
	}

	// Close writer to finalize the form
	if err := writer.Close(); err != nil {
		return nil, fmt.Errorf("failed to close multipart writer: %w", err)
	}

	// Make request
	resp, err := ie.client.doRequestWithReader(ctx, "POST", ie.client.buildPath("templates", "import"), &body, writer.FormDataContentType(), nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Check for error status
	if resp.StatusCode >= 400 {
		return nil, handleErrorResponse(resp)
	}

	// Decode response
	var result ImportTemplatesResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

// ExportAllTemplates exports all templates to YAML.
func (ie *ImportExportService) ExportAllTemplates(ctx context.Context, tag, names string) ([]byte, error) {
	queryParams := make(map[string]string)
	if tag != "" {
		queryParams["tag"] = tag
	}
	if names != "" {
		queryParams["names"] = names
	}
	return ie.client.doText(ctx, "GET", ie.client.buildPath("templates", "export"), queryParams)
}

File: sdk/python/src/liteboxd/services/import_export.py
"""Import/Export service for LiteBoxd SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..models import ImportResult, ImportStrategy

if TYPE_CHECKING:
    from ..client import Client


class ImportExportService:
    """Template import/export operations service."""

    def __init__(self, client: "Client") -> None:
        """Initialize import/export service."""
        self._client = client

    def import_templates(
        self,
        yaml_content: bytes | str,
        *,
        strategy: ImportStrategy = ImportStrategy.CREATE_OR_UPDATE,
        auto_prepull: bool = False,
    ) -> ImportResult:
        """
        Import templates from YAML.

        Args:
            yaml_content: YAML content (bytes or string)
            strategy: Import strategy (create-only, update-only, create-or-update)
            auto_prepull: Whether to auto-prepull images after import

        Returns:
            Import result with details of each template

        Raises:
            BadRequestError: Invalid YAML or request
        """
        if isinstance(yaml_content, str):
            yaml_content = yaml_content.encode("utf-8")

        files = {
            "file": ("templates.yaml", yaml_content, "application/x-yaml"),
        }
        data: dict[str, Any] = {
            "strategy": strategy.value,
        }
        if auto_prepull:
            data["prepull"] = "true"

        response = self._client._post("templates/import", data=data, files=files)
        return ImportResult.model_validate(response.json())

    def export_all(
        self,
        *,
        tag: str | None = None,
        names: list[str] | None = None,
    ) -> bytes:
        """
        Export all templates as YAML.

        Args:
            tag: Filter by tag
            names: Filter by template names

        Returns:
            YAML content as bytes
        """
        params: dict[str, Any] = {}
        if tag:
            params["tag"] = tag
        if names:
            params["names"] = ",".join(names)

        response = self._client._get("templates/export", params=params)
        return response.content
